File: ruby/difference-of-squares/difference_of_squares.rb
module BookKeeping
  VERSION = 4
end

class Squares
  def initialize(number)
    @range = 1..number
  end

  def square_of_sum
    @range.reduce(:+)**2
  end

  def sum_of_squares
    @range.reduce { |sum, num| sum + num**2 }
  end

  def difference
    square_of_sum - sum_of_squares
  end
end

File: go/difference-of-squares/difference_of_squares.go
package diffsquares

import (
	"math"
)

const square = 2.0

// SquareOfSums calculates the square of the sum of the first n numbers
func SquareOfSums(n int) int {
	return int(math.Pow(float64(n*(n+1)/2), square))
}

// SumOfSquares squares each of the first n numbers and adds them up
func SumOfSquares(n int) int {
	var sum float64
	for i := 1; i <= n; i++ {
		sum += math.Pow(float64(i), square)
	}
	return int(sum)
}

// Difference returns the square of sums minus the sum of the squares.
func Difference(n int) int {
	return SquareOfSums(n) - SumOfSquares(n)
}
